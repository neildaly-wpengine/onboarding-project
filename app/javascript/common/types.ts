export interface Article {
    title: string;
    content: string;
    createdAt: string;
    updatedAt: string;
    discardedAt?: string;
    archived: boolean;
    id: string;
    user: User;
    // todo remove this from backend serializer
    userID: number;
}

export interface User {
    id: string;
    firstName: string;
    lastName: string;
    initialsImageLink: string;
}