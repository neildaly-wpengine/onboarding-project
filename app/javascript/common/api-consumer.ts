import axios, { AxiosResponse } from "axios";
import humps from 'humps';
import {
    Article,
    CreateArticleBody,
    LoginBody,
    LogoutResponse,
    Registration,
    RegistrationBody,
    DeleteResponse
} from "./types";
const JSONAPIDeserializer = require("jsonapi-serializer").Deserializer;

class APIConsumer {
    async getAllArticles(): Promise<Article[]> {
        try {
            const response: AxiosResponse = await axios.get("/api/v1/articles");
            const articleData: Article[] = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e
        }
    }

    async getSpecificArticle(id: string): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.get(`/api/v1/articles/${id}`);
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    async registerNewUser(registrationBody: RegistrationBody): Promise<any> {
        try {
            const decamelizedKeys: RegistrationBody = humps.decamelizeKeys(registrationBody) as RegistrationBody;
            const response: Registration = await axios.post('/api/v1/registrations', decamelizedKeys,
                { withCredentials: true });

            return humps.camelizeKeys(response);
        } catch (e) {
            throw e;
        }
    }

    async destroySession(): Promise<LogoutResponse> {
        try {
            const response: AxiosResponse = await axios.delete('/api/v1/logout');

            return humps.camelizeKeys(response) as LogoutResponse;
        } catch (e) {
            throw e;
        }
    }

    async loginUser(loginBody: LoginBody): Promise<any> {
        try {
            const response: AxiosResponse = await axios.post('/api/v1/sessions', loginBody,
                { withCredentials: true });

            return humps.camelizeKeys(response);
        } catch (e) {
            throw e;
        }
    }

    async createArticle(articleBody: CreateArticleBody): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.post('/api/v1/articles',
                humps.decamelizeKeys(articleBody),
                { withCredentials: true });
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    async editArticle(articleBody: CreateArticleBody, articleID: string): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.patch(`/api/v1/articles/${articleID}`,
                humps.decamelizeKeys(articleBody),
                { withCredentials: true });
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    async deleteArticle(articleID: string): Promise<DeleteResponse> {
        try {
            const response: AxiosResponse = await axios.delete(`/api/v1/articles/${articleID}`,
                { withCredentials: true });

            return humps.camelizeKeys(response) as DeleteResponse;
        } catch (e) {
            throw e;
        }
    }

    async recoverArticle(articleID: string): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.post(`/api/v1/articles/${articleID}/recover`,
                { withCredentials: true });
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    private deserializeResponse = (response: AxiosResponse) => {
        const deserializerOptions = {
            keyForAttribute: "camelCase"
        }
        const deserializer = new JSONAPIDeserializer(deserializerOptions);
        return deserializer.deserialize(
            response.data,
            (_err: any, articlesResponseData: any) => {
                return articlesResponseData;
            }
        );
    }
}

export default APIConsumer;
