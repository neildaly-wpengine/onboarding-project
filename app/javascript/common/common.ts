import { User } from "./types";

export const createUserInitials = (user: User): string => {
    const userInitials: string = `${user.firstName.charAt(
        0
    )}${user.lastName.charAt(0)}`.toUpperCase();

    return userInitials;
}
