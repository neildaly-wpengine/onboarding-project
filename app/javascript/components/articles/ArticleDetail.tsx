import {
  Avatar,
  Box,
  CircularProgress,
  makeStyles,
  Typography,
} from "@material-ui/core";
import React, { useEffect, useState } from "react";
import {
  Article,
  ArticleDetailMatchParams,
  ConsumerProps,
} from "../../common/types";

const useStyles = makeStyles({
  boxHeader: {
    margin: 0,
    height: 450,
    padding: 0,
    overflow: "hidden",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  imageHeader: {
    objectFit: "cover",
    minWidth: "100%",
    minHeight: "100%",
    flexShrink: 0,
  },
  mainArea: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#fff",
  },
  mainHeader: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    textAlign: "center",
    marginTop: 15,
  },
  avatar: {
    alignSelf: "center",
  },
});

const ArticleDetail: React.FC<ArticleDetailMatchParams & ConsumerProps> = ({
  match,
  consumer,
}) => {
  const [article, setArticle] = useState<Article>();
  const classes = useStyles();

  useEffect(() => {
    let mounted: boolean = true;
    const fetchSpecificArticle = async () => {
      const articleID: string = match.params.id;
      const articleData = await consumer.getSpecificArticle(articleID);
      if (mounted) {
        setArticle(articleData);
      }
    };
    fetchSpecificArticle();

    return () => {
      mounted = false;
    };
  }, []);

  if (article === undefined) {
    return <CircularProgress data-testid="loading" />;
  }

  article.stockImage = `https://picsum.photos/1920/450?image=${article.id}`;

  return (
    <React.Fragment>
      <Box m={1} p={1} className={classes.boxHeader} width={1}>
        <img
          data-testid="specific-article-stock-image"
          src={article.stockImage}
          alt={article.title}
          className={classes.imageHeader}
        />
      </Box>
      <Box m={1} p={1} className={classes.mainArea} data-testid="resolved">
        <Box width="50%" className={classes.mainHeader}>
          <Typography variant="h2" gutterBottom>
            {article.title}
          </Typography>
          <Typography variant="h4" gutterBottom>
            {`${article.user.firstName} ${article.user.lastName}`}
          </Typography>
          <Avatar
            aria-label="user"
            src={article.user.initialsImageLink}
            className={classes.avatar}
          />
          <Typography variant="subtitle1" gutterBottom>
            {article.createdAt}
          </Typography>
        </Box>
        <Box width="50%" className={classes.mainHeader}>
          <Typography variant="body1" align="justify" gutterBottom>
            {article.content}
          </Typography>
        </Box>
      </Box>
    </React.Fragment>
  );
};

export default ArticleDetail;
