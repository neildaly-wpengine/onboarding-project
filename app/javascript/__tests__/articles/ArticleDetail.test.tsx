import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import { cleanup, render, RenderResult, waitFor } from "@testing-library/react";
import axios from "axios";
import { createLocation, createMemoryHistory } from "history";
import * as React from "react";
import { match } from "react-router";
import { ArticleDetailMatchParams } from "../../common/types";
import ArticleDetail from "../../components/articles/ArticleDetail";
import {
  consumer,
  jsonAPISpecificArticleResponseData,
  mockArticle,
} from "./__helpers__/test-data";

const history = createMemoryHistory();

const createRouteComponentProps = (): ArticleDetailMatchParams => {
  const path = `/articles/:id`;

  const match: match<{ id: string }> = {
    isExact: true,
    path,
    url: path.replace(":id", mockArticle.id),
    params: { id: mockArticle.id },
  };

  return {
    match: match,
    location: createLocation(match.url),
    history: history,
  };
};

const renderArticleDetail = (): RenderResult => {
  return render(
    <ArticleDetail consumer={consumer} {...createRouteComponentProps()} />
  );
};

afterEach(cleanup);

jest.mock("axios");

describe("<ArticleDetail />", () => {
  describe("Content", () => {
    it("renders correctly", async () => {
      const expectedResponse = {
        data: jsonAPISpecificArticleResponseData,
      };

      (axios.get as jest.Mock).mockImplementationOnce(() =>
        Promise.resolve(expectedResponse)
      );

      const { container, getByTestId } = renderArticleDetail();
      expect(getByTestId("loading")).toHaveAttribute("role", "progressbar");

      const resolvedData = await waitFor(() => getByTestId("resolved"));

      expect(container).toMatchSnapshot();
      expect(getByTestId("specific-article-stock-image")).toHaveAttribute(
        "src",
        `https://picsum.photos/1920/450?image=${mockArticle.id}`
      );
      expect(axios.get).toHaveBeenCalledWith(
        `/api/v1/articles/${mockArticle.id}`
      );
      expect(resolvedData).toBeInTheDocument();
      expect(resolvedData).not.toBeEmptyDOMElement();
    });
  });
});
