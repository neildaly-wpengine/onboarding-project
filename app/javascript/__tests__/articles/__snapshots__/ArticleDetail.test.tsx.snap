// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleDetail /> Content renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-6 makeStyles-boxHeader-1"
  >
    <img
      alt="Computer Games Development"
      class="makeStyles-imageHeader-2"
      data-testid="specific-article-stock-image"
      src="https://picsum.photos/1920/450?image=1"
    />
  </div>
  <div
    class="MuiBox-root MuiBox-root-7 makeStyles-mainArea-3"
    data-testid="resolved"
  >
    <div
      class="MuiBox-root MuiBox-root-8 makeStyles-mainHeader-4"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom"
      >
        Computer Games Development
      </h2>
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom"
      >
        Harry Smith
      </h4>
      <div
        aria-label="user"
        class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-5"
      >
        <img
          class="MuiAvatar-img"
          src="https://eu.ui-avatars.com/api/?background=fff&color=000&name=Harry+Smith"
        />
      </div>
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom"
      >
        2020-06-25T10:07:42.357Z
      </h6>
    </div>
    <div
      class="MuiBox-root MuiBox-root-9 makeStyles-mainHeader-4"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiTypography-alignJustify"
      >
        Lorem ipsum dolor sit 
      </p>
    </div>
  </div>
</div>
`;
